### PR TITLE
Bring LDAP user attribute updates ("sync") to the background

### PR DESCRIPTION
**Pull request by blizzz**
Needed for #6982, but pretty nice in general

### CODE REVIEW

**Comment by blizzz** [comment]
I am going to extend sync unit tests (did a lot of smoke tests today), but for I had enough today… ;) Otherwise I am happy about reviews @nextcloud/ldap

**Comment by blizzz** [comment]
Latest measuring showed >85% performance increase in the scenario:

* fetch 500 users, offset 1000,  search term "ha" against 3 attributes
* users fetched are already mapped, background job is set to cron
* redis is flushed before each run
* on CLI using ``time sudo -u http ./occ ldap:search --limit=500 --offset=1000 "ha"``, taking the real value.
* n=10 for each master and branch

With current master average runtime was 15.2s, this branch does  it in 1.8s. 

This does not apply to users that are new to Nextcloud, aka are being fetched for the first time. Fortunately this happens once. A background rotates through the servers and fetches a batch, also it adjusts the interval it runs depending on the number of known users, within a minimum and maximum interval. And also it runs earliest in one hour after a config change to avoid running to early when it might not be complete.

File: apps/user_ldap/lib/Access.php [hunk 1]
```diff
@@ -86,19 +87,27 @@ class Access extends LDAPUtility implements IUserTools {
...
-		Manager $userManager, Helper $helper) {
+	/** @var IServerContainer */
+	private $c;
+
+	public function __construct(
+		Connection $connection,
+		ILDAPWrapper $ldap,
+		Manager $userManager,
+		Helper $helper,
+		IServerContainer $c
```
Comment (LukasReschke) [inline apps/user_ldap/lib/Access.php]: Why do we inject the server container here? This kinda hides the real dependency which would be `IConfig` as far as I can see :)

File: apps/user_ldap/lib/Access.php [hunk 2]
```diff
@@ -511,12 +520,16 @@ public function dn2username($fdn, $ldapName = null) {
 
 	/**
 	 * returns an internal Nextcloud name for the given LDAP DN, false on DN outside of search DN
-	 * @param string $dn the dn of the user object
+	 *
+	 * @param string $fdn the dn of the user object
 	 * @param string $ldapName optional, the display name of the object
```
Comment (LukasReschke) [inline apps/user_ldap/lib/Access.php]: string|null? Since you also changed those below :)

File: apps/user_ldap/lib/Jobs/Sync.php [hunk 3]
```diff
@@ -0,0 +1,330 @@
...
+use OCA\User_LDAP\Mapping\UserMapping;
+use OCA\User_LDAP\User\Manager;
+use OCP\Image;
+use OCP\IServerContainer;
+
+class Sync extends TimedJob {
+	const MAX_INTERVAL = 12 * 60 * 60; // 12h
+	const MIN_INTERVAL = 30 * 60; // 30min
+	/** @var IServerContainer */
+	protected $c;
```
Comment (LukasReschke) [inline apps/user_ldap/lib/Jobs/Sync.php]: Also here I'm not really a fan of hiding the dependency away :)

**Review by LukasReschke — DISMISSED** [review]
Code looks good. Just some minor nitpicks :)

Didn't do a functional test yet. Will try tomorrow.

File: apps/user_ldap/lib/Access.php (hunk 1 above)
Comment (blizzz) [reply apps/user_ldap/lib/Access.php]: For the newly inserted classes, yes, the deprecated login stuff could be replaced as well. At some point. Unless I overhaul it generally. The thought came to my mind as well however.

File: apps/user_ldap/lib/Jobs/Sync.php (hunk 3 above)
Comment (blizzz) [reply apps/user_ldap/lib/Jobs/Sync.php]: More complex here, but since it's not really instantiated anywhere else but Tests no problem to adjust this too.

**Comment by blizzz** [comment]
@LukasReschke  thanks! your comments are addressed

**Comment by blizzz** [comment]
rebased, due to conflicts. 

more reviews, please please please 🐱

**Review by skjnldsv — APPROVED** [review]
Works perfectly fine!!

**Comment by blizzz** [comment]
another one? @nextcloud/ldap @LukasReschke @rullzer @MorrisJobke ?

**Review by LukasReschke — APPROVED** [review]
🙏 – stuff still seems to work.

**Comment by blizzz** [comment]
Thank you @skjnldsv and @LukasReschke 🙇

**Pull request merged**